src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -203,69 +251,241 @@ public object[] ArgumentList
+            switch (ParameterSetName)
+            {
+                case ForEachObjectCommand.ParallelParameterSet:
+                    StopParallelProcessing();
+                    break;
+            }
+        }
+
+        #endregion
+
+        #region IDisposable
+
+        /// <summary>
+        /// Dispose cmdlet instance.
+        /// </summary>
+        public void Dispose()
+        {
+            // Ensure all parallel task objects are disposed
+            if (_taskTimer != null)
+            {
+                _taskTimer.Dispose();
+            }
+            if (_taskDataStreamWriter != null)
+            {
+                _taskDataStreamWriter.Dispose();
+            }
+            if (_taskPool != null)
+            {
+                _taskPool.Dispose();
+            }
+        }
+
+        #endregion
+        
+        #region Private Methods
+
+        #region PSTasks
+
+        private PSTaskPool _taskPool;
+        private PSTaskDataStreamWriter _taskDataStreamWriter;
+        private Dictionary<string, object> _usingValuesMap;
+        private System.Threading.Timer _taskTimer;
+        private PSTaskJob _taskJob;
+
+        private void InitParallelParameterSet()
+        {
+            bool allowUsingExpression = this.Context.SessionState.LanguageMode != PSLanguageMode.NoLanguage;
+            _usingValuesMap = ScriptBlockToPowerShellConverter.GetUsingValuesAsDictionary(Parallel, allowUsingExpression, this.Context, null);
+            
+            // Validate using values map
+            foreach (var item in _usingValuesMap.Values)
+            {
+                if (item is ScriptBlock)
                 {
-                    whatIf = (SwitchParameter)argument;
+                    this.ThrowTerminatingError(
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelUsingVariableCannotBeScriptBlock),
+                            "ParallelUsingVariableCannotBeScriptBlock",
+                            ErrorCategory.InvalidType,
+                            this));
                 }
+            }
 
-                if (psBoundParameters.TryGetValue("confirm", out argument))
+            if (AsJob)
+            {
+                if (MyInvocation.BoundParameters.ContainsKey(nameof(TimeoutSeconds)))
                 {
-                    confirm = (SwitchParameter)argument;
+                    this.ThrowTerminatingError(
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelCannotUseTimeoutWithJob),
+                            "ParallelCannotUseTimeoutWithJob",
+                            ErrorCategory.InvalidOperation,
+                            this));
                 }
 
-                if (whatIf || confirm)
+                _taskJob = new PSTaskJob(
+                    Parallel.ToString(),
+                    ThrottleLimit);
+            }
+            else
```

Comment:
Since this is the end of the method, it might be worth returning early rather than using `else`